pennylane_lightning/src/util/Cache.hpp

```diff
@@ -0,0 +1,196 @@
+// Copyright 2021 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+/**
+ * @file
+ * Least Recently Updated (LRU) cache policy class implementation.
+ */
+
+#pragma once
+
+/* ======================================================= */
+/* LRU Library Header */
+/* ======================================================= */
+
+#include <list>
+#include <unordered_map>
+#include <vector>
+
+#include <limits>
+#include <utility>
+
+namespace Pennylane {
+
+namespace Util {
+
+// Boost implementation of a hash combine:
+// https://www.boost.org/doc/libs/1_35_0/doc/html/boost/hash_combine_id241013.html
+struct hash_function {
+    std::size_t
+    operator()(const std::pair<std::vector<size_t>, size_t> &key) const {
+        std::size_t combined_hash_value = 0;
+
+        for (auto &term : key.first) {
+            combined_hash_value ^= std::hash<size_t>()(term) + 0x9e3779b9 +
+                                   (combined_hash_value << 6) +
+                                   (combined_hash_value >> 2);
+        };
+        combined_hash_value ^= std::hash<size_t>()(key.second) + 0x9e3779b9 +
+                               (combined_hash_value << 6) +
+                               (combined_hash_value >> 2);
+        return combined_hash_value;
+    }
+};
+
+/**
+ * @brief Least Recently Updated (LRU) cache policy class.
+ *
+ * This class define a container to store previously calculated indices.
+ * Data is stored up to a maximum cache size, defined by the class constructor.
+ * The key for accessing the cache storage is a vector and a scalar
+ * representing respectively, the wires and number of qubits involved in the
+ * calculation.
+ *
```

Comment:
To help with the documentation generation, the comment here should have the template parameters as `@tparam` entries (e.g see [here](https://github.com/PennyLaneAI/pennylane-lightning/blob/238f4d05ae6b2aa3d06fd3e77113dc847650d2a6/pennylane_lightning/src/simulator/StateVector.hpp#L66) for how this is done with the StateVector class).